src/monotonic_path_preview/mod.rs

```diff
@@ -0,0 +1,911 @@
+    /// [`read_link()`]: https://doc.rust-lang.org/stable/std/fs/fn.read_link.html
+    ///
+    /// # Platform-specific behaviour
+    ///
+    /// On Windows, this method correctly handles partially-absolute paths like
+    /// `D:foo\bar.txt` that are relative to a path other than the current
+    /// working directory.
+    ///
+    /// On Windows, the resulting path uses [extended-length path syntax], so
+    /// it may confuse other applications not designed to handle such paths.
+    /// For example, if you pass such a path on another application's command
+    /// line, or write it to a configuration ile.
+    ///
+    /// [extended-length path syntax]: https://msdn.microsoft.com/en-us/library/windows/desktop/aa365247(v=vs.85).aspx#maxpath
+    ///
+    /// # Errors
+    ///
+    /// Returns [`Error::IoError`] if the head of the given path cannot be
+    /// canonicalized. For example, if the process' current working directory
+    /// has been deleted, or the given path includes a syntactically invalid
+    /// prefix.
+    ///
+    /// The same variant is returned if a problem is encountered while checking
+    /// if a given path is a symlink, or while trying to read a symlink. For
+    /// example, if the current user does not have permission to read the
+    /// directory containing it, or the path is on a network-mounted filesystem
+    /// that stopped responding.
+    ///
+    /// Returns [`Error::SymlinkLoops`] if resolving a symlink takes us back
+    /// to a previously-resolved symlink. For example, if `/example/path/a`
+    /// is a symlink to `/example/path/b`, and `b` is a symlink back to `a`,
+    /// then giving this method a path like `/example/path/a/../c` will return
+    /// this error. It's like the POSIX `ELOOP` error, but cross-platform.
+    ///
+    /// Note that "does not exist" is *not* a fatal error for this function;
+    /// path components that do not exist by definition are not symlinks, and
+    /// are treated the same way as every other component that is not a symlink.
+    ///
+    /// [`Error::IoError`]: struct.Error.html#variant.IoError
+    /// [`Error::SymlinkLoops`]: struct.Error.html#variant.SymlinkLoops
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # fn example() -> Result<(), Error> {
+    /// let real_current_directory = Absolute::new(std::env::current_dir()?)?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn new<P: AsRef<path::Path>>(path: P) -> Result<Absolute, Error> {
+        // This function steps through the components of path, checking each one
+        // for monotonicity and building up a proper monotonic path in `res`.
+        // `tail` contains the path components we've yet to check and clean up.
+        let (mut res, mut tail) = split_head_and_tail(path)?;
+
+        let mut seen_paths = collections::BTreeSet::new();
+
+        // While there's still components left to check...
+        while tail.len() > 0 {
```

Comment:
Please break up the logic in this loop into separate functions. It's very hard to read.